storage/internal/file/store.go

```diff
@@ -0,0 +1,462 @@
+	// decrypt if necessary
+	if s.withEncryption {
+		data, err = s.encryptor.Decrypt(data)
+		if err != nil {
+			return false, nil, err
+		}
+	}
+
+	return true, data, errors.Wrapf(s.codec.Unmarshal(data, v), "failed to unmarshal data:%q", string(data))
+}
+
+func (s *fileStore) Exist(k string) (bool, error) {
+	filePath := s.buildFilePath(k, true)
+
+	_, err := os.Stat(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+func (s *fileStore) Keys(k string) ([]string, error) {
+	filePath := s.buildFilePath(k, false)
+
+	files, err := ioutil.ReadDir(filePath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			return nil, nil
+		}
+		return nil, err
+	}
+
+	result := make([]string, 0)
+	for _, file := range files {
+		fileName := file.Name()
+		// return the whole key path without specific extension
+		result = append(result, path.Join(k, strings.TrimSuffix(fileName, "."+s.filenameExtension)))
+	}
+
+	return collections.RemoveDuplicates(result), nil
+}
+
+func (s *fileStore) Delete(ctx context.Context, k string, recursive bool) error {
+	if err := utils.CheckKey(k); err != nil {
+		return err
+	}
+
+	// Handle cache
+	if s.withCache {
+		if err := s.clearCache(ctx, k, recursive); err != nil {
+			return err
+		}
+	}
+
+	var err error
+
+	// Try to delete a single file in all cases
+	filePath := s.buildFilePath(k, true)
+	// Prepare file lock.
+	lock := s.prepareFileLock(filePath)
+
+	// File lock and file handling.
+	lock.Lock()
+	defer lock.Unlock()
+
+	err = os.Remove(filePath)
+
+	// Try to delete a directory if recursive is true
+	if recursive {
+		// Remove the whole directory
+		err = os.RemoveAll(s.buildFilePath(k, false))
+	}
+
+	if os.IsNotExist(err) {
+		return nil
+	}
+	return err
+}
+
+func (s *fileStore) clearCache(ctx context.Context, k string, recursive bool) error {
+	s.cache.Del(k)
+
+	if !recursive {
+		return nil
+	}
+
+	// Delete sub-keys
+	keys, err := s.Keys(k)
+	if err != nil {
+		return err
+	}
+	errGroup, ctx := errgroup.WithContext(ctx)
+	for _, key := range keys {
+		key := key
+		errGroup.Go(func() error {
+			return s.clearCache(ctx, key, true)
+		})
+	}
+
+	return errGroup.Wait()
+}
+
+func (s *fileStore) GetLastModifyIndex(k string) (uint64, error) {
+	// key can be directory or file, ie with or without extension
+	// let's try first without extension as it can be the most current case
+	rootPath := s.buildFilePath(k, false)
+	fInfo, err := os.Stat(rootPath)
+
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return 0, errors.Wrapf(err, "failed to get last index for key:%q", k)
+		}
+		// not a directory, let's try a file with extension
+		fp := s.buildFilePath(k, true)
+		fInfo, err = os.Stat(fp)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return 0, errors.Wrapf(err, "failed to get last index for key:%q", k)
+			}
+			// File not found : the key doesn't exist
+			return 0, nil
+		}
+		return uint64(fInfo.ModTime().UnixNano()), nil
```

Comment:
Yes but what if I `touch` the file?

Reply:
The modifyIndex is updated as if you modify a key in Consul UI